File: src/main/java/leetcode/hot100/LeetCode647.java
package leetcode.hot100;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/5 22:20
 */
public class LeetCode647 {

    //https://leetcode.com/problems/palindromic-substrings/discuss/2369314/Java-Solution-or-Faster-than-99.99-or-O(n2)

    public int countSubstrings(String s) {
        char[] c = s.toCharArray();
        int n = c.length, ans = 0;
        for (int i = 0; i < n; i++) {
            // odd-length strings centered at c[i]
            int l = i, r = i;
            while (l >= 0 && r < n && c[l] == c[r]) {
                ans++;
                l--;
                r++;
            }
            //even-length strings centered at c[i] and c[i+1]
            l = i;
            r = i + 1;
            while (l >= 0 && r < n && c[l] == c[r]) {
                ans++;
                l--;
                r++;
            }
        }
        return ans;
    }
}

File: go/src/leetcode/hot100/LeetCode647.go
package hot100

func countSubstrings(s string) int {
	if len(s) == 0 {
		return 0
	}
	dp := make([][]int, len(s))
	for i := 0; i < len(s); i++ {
		dp[i] = make([]int, len(s))
	}
	var res int
	for i := len(s) - 1; i >= 0; i-- {
		//为什么不从0开始? 因为dp[i][j] j>=i 才有意义
		//所以最重要的还是状态转移方程
		//状态转移方程能决定遍历的顺序，以及遍历的开始点和结束点
		for j := i; j < len(s); j++ {
			if i == j {
				dp[i][j] = 1
			} else {
				if s[i] == s[j] {
					if dp[i+1][j-1] > 0 || (j-i == 1) {
						dp[i][j] = dp[i+1][j-1] + 2
					}
				}
			}
			if dp[i][j] > 0 {
				res++
			}
		}
	}
	return res
}
